src/hotspot/share/classfile/loaderConstraints.cpp

```diff
@@ -27,80 +27,158 @@
 #include "classfile/classLoaderDataGraph.hpp"
 #include "classfile/dictionary.hpp"
 #include "classfile/loaderConstraints.hpp"
+#include "classfile/placeholders.hpp"
 #include "logging/log.hpp"
 #include "memory/resourceArea.hpp"
 #include "oops/klass.inline.hpp"
 #include "oops/oop.inline.hpp"
 #include "runtime/handles.inline.hpp"
+#include "runtime/mutexLocker.hpp"
 #include "runtime/safepoint.hpp"
-#include "utilities/hashtable.inline.hpp"
+#include "utilities/resourceHash.hpp"
+
+// Implementation Classes for Loader Constraints
+
+class LoaderConstraint : public CHeapObj<mtClass> {
+  InstanceKlass*         _klass;
+  // Loader constraints enforce correct linking behavior.
+  // Thus, it really operates on ClassLoaderData which represents linking domain,
+  // not class loaders.
+  GrowableArray<ClassLoaderData*>*  _loaders;                // initiating loaders
+ public:
+  LoaderConstraint(InstanceKlass* klass, oop class_loader1, oop class_loader2) :
+     _klass(klass) {
+    _loaders = new (ResourceObj::C_HEAP, mtClass) GrowableArray<ClassLoaderData*>(10, mtClass);
+    set_loader(class_loader1);
+    set_loader(class_loader2);
+  }
 
-void LoaderConstraintEntry::set_loader(int i, oop p) {
-  set_loader_data(i, ClassLoaderData::class_loader_data(p));
-}
+  LoaderConstraint() { delete _loaders; }
 
-LoaderConstraintTable::LoaderConstraintTable(int table_size)
-  : Hashtable<InstanceKlass*, mtClass>(table_size, sizeof(LoaderConstraintEntry)) {};
+  InstanceKlass* klass() const     { return _klass; }
+  void set_klass(InstanceKlass* k) { _klass = k; }
 
+  void extend_loader_constraint(Symbol* class_name, Handle loader, InstanceKlass* klass);
 
-LoaderConstraintEntry* LoaderConstraintTable::new_entry(
-                                 unsigned int hash, Symbol* name,
-                                 InstanceKlass* klass, int num_loaders,
-                                 int max_loaders) {
-  LoaderConstraintEntry* entry;
-  entry = (LoaderConstraintEntry*)Hashtable<InstanceKlass*, mtClass>::new_entry(hash, klass);
-  entry->set_name(name);
-  entry->set_num_loaders(num_loaders);
-  entry->set_max_loaders(max_loaders);
-  return entry;
-}
+  int num_loaders() const { return _loaders->length(); }
+  ClassLoaderData* loader_data(int i) { return _loaders->at(i); }
+  void set_loader_data(ClassLoaderData* p) { _loaders->push(p); }
+
+  void remove_loader_at(int n) {
+    assert(_loaders->at(n)->is_unloading(), "should be unloading");
+    _loaders->remove_at(n);
+  }
+
+  // convenience
+  void set_loader(oop p) {
+    _loaders->push(ClassLoaderData::class_loader_data(p));
+  }
+};
+
+class LoaderConstraintEntry : public StackObj {      // copied into hashtable as value
```

Comment:
StackObj does not seem appropriate for something being stored in a hashtable.

Reply:
Yes, I'd been using StackObj for the copy to put into the resource hashtable, removed.